Src/Newtonsoft.Json/JsonValidatingReader.cs

```diff
@@ -710,15 +710,14 @@ private void ValidateEndObject(JsonSchemaModel schema)
 
             Dictionary<string, bool> requiredProperties = _currentScope.RequiredProperties;
 
-            if (requiredProperties != null)
+            if (requiredProperties != null && requiredProperties.Values.Any(v => !v))
```

Comment:
This benefits the non-throwing branch at the cost of more work to the throwing, which is generally a win for most cases.